docs/proposal-platforms.md

```diff
@@ -0,0 +1,298 @@
+```ini
+[platforms]
+    [[hpcl[12]-bg]]
+        hosts = ("-bg$","") # or "hosts = s/-bg$//" ?
+        retrieve job logs = True
+        batch system = background
+```
+
+## Example platform usage
+
+These are examples of how the platform may be assigned in the `[runtime]`
+section for a job. They refer to the example platforms shown above.
+
+`platform = desktop01`
+* The simplest example. The job will be run on the host `desktop01`.
+
+`platform = special`
+* There is no platform named `special` defined so this will fail validation.
+
+`platform = hpc`
+* The host used will be chosen at random.
+
+`platform = hpc-bg`
+* The platform used will be chosen at random ("hpcl1" or "hpcl2").
+
+`platform = $(rose host-select linux)`
+* No checking can be performed at validation so if the command does not return a
+  valid platform this will result in a submit failure.
+
+## Deprecation method
+
+* You can't mix old with new. If `platform` is defined for **any** task then any
+  settings in the `[remote]` or `[job]` sections of **any** task result in a
+  validation failure.
+
+* Any settings in the `[remote]` or `[job]` sections which have equivalent
+  settings in the new runtime section are converted using the normal deprecate
+  method. e.g. `[job] execution time limit`.
+
+* The remaining `[remote]` & `[job]` settings are converted by searched for a
+  matching platform. If there is none the conversion fails resulting in an
+  error.
+
+* For fixed remote hosts the conversion happens at load time and the lack of a
+  matching platform results in a validation failure.
+
+* For remote hosts defined as a command or an environment variable the
+  remaining `[remote]` & `[job]` settings are stored and the conversion happens
+  at job submission. In this case the lack of a matching platform results in a
+  submission failure.
+
+* The `[remote] host` setting is matched first by checking the platforms
+  `hosts` setting. If this is not defined then the match is made against
+  the valid platform names.
+  * We may need special logic so that if `[remote] host` is set to the workflow
+    server hostname then this is converted to `localhost` before matching
+    occurs.
+
+## Deprecation examples
+
+```ini
+[runtime]
+    [[alpha]]
+        [[[remote]]]
+            host = localhost
+        [[[job]]]
+            batch system = background
+        # => platform = localhost (set at load time)
+
+    [[beta]]
+        [[[job]]]
+            batch system = slurm
+        # => platform = sugar (set at load time)
+
+    [[gamma]]
+        [[[remote]]]
+            host = $(rose host-select hpc)
+            # assuming this returns "hpcl1" or "hpcl2"
+        [[[job]]]
+            batch system = pbs
+        # => platform = hpc (set at job submission time)
+
+    [[delta]]
+        [[[remote]]]
+            host = hpcl1
+        [[[job]]]
+            batch system = background
+        # => platform = hpcl1-bg (set at load time)
+```
```

Comment:
I'd like an example of settings that will fail because the items in `[[[remote]]]` and `[[[job]]]` ≠ any platform. e.g.
```config
# Will fail loudly
[[epsilon]]
    [[[remote]]]
        host = desktop99
    [[[job]]]
         batch system = pbs
```

Reply:
Further examples added